GHC 9.2 build support

avieth opened the pull request.

avieth committed:
make it build on ghc 9.2.1

Not backwards compatible

avieth committed:
restore hoverInfo

Not sure if it's actually correct, but it builds.

masaeedu commented on lib/GhcideSteal.hs:
```diff
@@ -10,24 +10,32 @@ import Control.Monad.Trans.Class
 import HieDb hiding (pointCommand)
-import HieTypes
-import HieUtils
+--import HieTypes
+--import HieUtils
```
```suggestion
```

masaeedu commented on lib/GhcideSteal.hs:
```diff
@@ -10,24 +10,32 @@ import Control.Monad.Trans.Class
-import SrcLoc
 import System.FilePath ((</>))
 
+mapMaybeM :: Monad m => (a -> m (Maybe b)) -> ([a] -> m [b])
```
Would this do the same thing as `\f -> fmap catMaybes . traverse f`?

avieth replied:
Yes I think so.

masaeedu commented:
Hi Alex! Thank you for setting this up, this should be super handy.

I have a busy couple of days ahead of me, but I'll try to take a look right after. I know this asking for more work on top of what you've already done, but would you mind sprinkling in some GitHub comments to summarize what's being changed in broad strokes?

This will make it much quicker to review and can serve as a reference if I have to come back to understand what's going on here.

avieth commented on lib/Utils.hs:
```diff
@@ -44,11 +44,11 @@ ekbind k f cb = k $ either (cb . Left) (($ cb) . f)
 
 kuncodensity :: Monad m => (forall r. K (m r) (m r) a) -> m a
-kuncodensity = ($ pure)
+kuncodensity k = k pure
```
These definitions are no longer well-typed in GHC 9.2 because of simplified subsumption. Patched versions are in a sense "the same thing" but pass GHC's new checker.

avieth commented on lib/GhcideSteal.hs:
```diff
@@ -136,25 +149,50 @@ locationsAtPoint ::
+-- Copied from nodeInfo, which is specialized to Type, because it uses
+-- nonDetCmpType for the ordering... but TypeIndex is an Int. FIXME does the
+-- ordering here actually make sense? Is this definition complete trash? Could
+-- be.
```
This definition `nodeInfo'` needs a review.

avieth reviewed (commented):
It's actually been a while since I made these changes. The bulk of it was necessitated by big changes to the GHC API library: check it out at [8.10.7](https://hackage.haskell.org/package/ghc-8.10.7) versus [9.2.1](https://hackage.haskell.org/package/ghc-9.2.1).

The use of `HieTypes.nodeInfo` in` GhcideSteal.hs`, for example, doesn't make any sense anymore and is not even well-typed.

avieth committed:
add generated cabal file

josephsumabat committed:
Resolve module name from imports if possible

sometimes `modinfoSrcFile` cannot resolve the file name
since of a module despite the file being indexed by hiedb
causing jump to definition to fail even though we have the necessary
information (I believe this can happen if it's a seperate build target
eg test vs library). We can use `modInfoName` to resolve the module and
which fixes the jump to definition

avieth committed:
Merge pull request #1 from josephsumabat/joseph/ghc92/getModuleName

Joseph/ghc92/get module name
